runners/direct-java/src/main/java/org/apache/beam/runners/direct/TestStreamEvaluatorFactory.java

```diff
@@ -68,35 +83,60 @@
   @Override
   public void cleanup() throws Exception {}
 
+  /**
+   * Returns the evaluator for the provided application of {@link TestStream}, or null if it is
+   * already in use.
+   *
+   * <p>The documented behavior of {@link TestStream} requires the output of one event to travel
+   * completely through the pipeline before any additional event, so additional instances that
+   * have a separate collection of events cannot be created.
+   */
   private <InputT, OutputT> TransformEvaluator<? super InputT> createEvaluator(
       AppliedPTransform<PBegin, PCollection<OutputT>, TestStream<OutputT>> application,
       EvaluationContext evaluationContext) {
-    if (evaluator.get() == null) {
-      Evaluator<OutputT> createdEvaluator = new Evaluator<>(application, evaluationContext, inUse);
-      evaluator.compareAndSet(null, createdEvaluator);
-    }
-    if (inUse.compareAndSet(false, true)) {
-      return evaluator.get();
-    } else {
-      return null;
+    // Replace any existing value with absent, and get the existing value (atomically); ensures
```

Comment:
Suggestion:

```
public static class ConcurrentSingleUseInstanceCache<K, V> {
  public ConcurrentSingleUseInstanceCache(Function<K, V> createInstance) { ... }
  public @Nullable V tryAcquire(K key) { ... };
  public void release(K key, V value) { ... }
}
```

That will isolate the use pattern of the concurrency primitives behind a clear API, and also make it easier to test. Someone once told me to write some multi-threaded tests for this kind of thing.